File: 프로그래머스/lv3/92343. 양과 늑대/양과 늑대.java
import java.util.*;

class Solution {
    int[][][] table = new int[1 << 17][18][18];

    public int solution(int[] info, int[][] edges) {
        return solve(1, 1, 0, info, edges);
    }

    int solve(int state, int sheep, int wolf, int[] info, int[][] edges) {
        if (table[state][sheep][wolf] > 0) return table[state][sheep][wolf];

        table[state][sheep][wolf] = sheep;

        for (int[] edge : edges) {
            if ((state & (1 << edge[0])) != 0 && (state & (1 << edge[1])) == 0) {
                if (info[edge[1]] == 0)
                    table[state][sheep][wolf] = Math.max(table[state][sheep][wolf], solve(state | (1 << edge[1]), sheep + 1, wolf, info, edges));
                else if (sheep > wolf + 1)
                    table[state][sheep][wolf] = Math.max(table[state][sheep][wolf], solve(state | (1 << edge[1]), sheep, wolf + 1, info, edges));
            }
        }

        return table[state][sheep][wolf];
    }
}

File: 프로그래머스/lv3/92343. 양과 늑대/양과 늑대.cpp
#include <iostream>
#include <string>
#include <cstring>
#include <vector>
using namespace std;

int n;
int table[1 << 17][18][18];
vector<int> child[17];

int solve(int state, int sheep, int wolf, vector<int> &info) {
  int &ret = table[state][sheep][wolf];
  if (ret != -1) return ret;

  ret = sheep;

  for (int i = 0; i < n; i++) {
    if (state & (1 << i)) {
      for (auto iter : child[i]) {
        if (!(state & (1 << iter)) && sheep + !info[iter] > wolf + info[iter])
          ret = max(ret, solve(state | (1 << iter), sheep + !info[iter], wolf + info[iter], info));
      }
    }
  }
  return ret;
}

int solution(vector<int> info, vector<vector<int>> edges) {
  memset(table, -1, sizeof(table));

  n = info.size();
  for (auto iter : edges) {
    child[iter[0]].push_back(iter[1]);
  }

  return solve(1, 1, 0, info);
}